File: exo4.1.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

/*Écrire un programme qui crée deux processus. Le processus père
ouvre un ﬁchier texte en lecture. On suppose que le ﬁchier est composé de mots formés de
caractères alphabétiques séparés par des espaces. Le processus ﬁls saisit un mot au clavier. Le
processus père recherche le mot dans le ﬁchier, et transmet au ﬁls la valeur 1 si le mot est dans
le ﬁchier, et 0 sinon. Le ﬁls aﬃche le résultat.*/
int main () {

    int tube[2];
    char motSaisie[64], motFichier[64];

    pipe (tube);
    pid_t pid = fork ();

    if (pid > 0) {
        close (tube[0]);
        printf ("Entrez le mot a rechercher : ");
        scanf ("%s", motSaisie);

        FILE* fp = fopen ("text.txt", "r");
        int isThere = 0;
        while (fscanf (fp, "%s", motFichier) != EOF) {
            if (strcmp (motSaisie, motFichier) == 0) {
                isThere = 1;
                break;
            }
        }

        write (tube[1], &isThere, sizeof (int));
        fclose (fp);
        close (tube[1]);
    } else {
        close (tube[1]);

        int resultat;
        read (tube[0], &resultat, sizeof (int));

        printf ("Mot trouve ? %d\n", resultat);
        close (tube[0]);
    }

    return 0;
}

File: exo4.1.py
import os
import sys

# Programme qui crée deux processus.
# Le père ouvre un fichier texte, le fils saisit un mot au clavier.
# Le père recherche le mot dans le fichier et transmet le résultat au fils.

def main():
    r, w = os.pipe()
    pid = os.fork()

    if pid > 0:  # Processus père
        os.close(r)
        mot_saisie = input("Entrez le mot a rechercher : ")

        is_there = 0
        try:
            with open("text.txt", "r") as fp:
                for ligne in fp:
                    for mot_fichier in ligne.split():
                        if mot_saisie == mot_fichier:
                            is_there = 1
                            break
                    if is_there:
                        break
        except FileNotFoundError:
            print("Fichier text.txt non trouvé.")

        os.write(w, is_there.to_bytes(4, 'little'))
        os.close(w)
        os.wait()  # Attendre la fin du fils
    else:  # Processus fils
        os.close(w)
        resultat_bytes = os.read(r, 4)
        resultat = int.from_bytes(resultat_bytes, 'little')
        print(f"Mot trouve ? {resultat}")
        os.close(r)

if __name__ == "__main__":
    main()
